NetP Status View Header

graeme opened the pull request:
<!--
Note: This checklist is a reminder of our shared engineering expectations. Feel free to change it, although assigning a GitHub reviewer and the items in bold are required.

⚠️ If you're an external contributor, please file an issue first before working on a PR, as we can't guarantee that we will accept your changes if they haven't been discussed ahead of time. Thanks!
-->

Task/Issue URL: https://app.asana.com/0/0/1205084446087078/f

**Description**:

This is the first design implementation task in the overarching Network Protection Status View task and covers just the header view (that is, the views above the toggle view). I’m delaying on making this look good on iOS 14 for now as it will take some extra thought. But I am checking in with the Apple team and have created a follow-up task to fix iOS 14 at the end of the project if need be.

**Steps to test this PR**:
1. (If you haven’t followed the steps in [Network Protection iOS Smoke Tests: Invite Code flow](https://app.asana.com/0/0/1205201898139908) already, do so.)
2. **Check that the status view header (the views above, not including, the toggle view) matches [the designs](https://www.figma.com/file/zuCBX2UpJpS8CgtcmpU1hf/Network-Protection-for-iOS?type=design&node-id=252-21943&mode=design&t=Z7gEv95kIs3uMn7e-0) disconnected state** 
3. Switch the toggle to activate Network Protection.
4. **Check that the status view header again matches the designs but for the connected state**

**Copy Testing**:

* [ ] Use of correct apostrophes in new copy, ie `’` rather than `’`

**Orientation Testing**:

* [ ] Portrait
* [ ] Landscape

**Device Testing**:

* [ ] iPhone SE (1st Gen)
* [x] iPhone 8
* [ ] iPhone X
* [ ] iPhone 14 Pro
* [ ] iPad

**OS Testing**:

* [ ] iOS 14
* [x] iOS 15
* [x] iOS 16

**Theme Testing**:

* [x] Light theme
* [x] Dark theme

—
###### Internal references:
[Software Engineering Expectations](https://app.asana.com/0/59792373528535/199064865822552)
[Technical Design Template](https://app.asana.com/0/59792373528535/184709971311943)

graeme committed:
Add status view image assets

graeme committed:
Point to BSK branch with observer improvements

graeme committed:
Unembed PacketTunnelProvider

graeme committed:
Point to updated BSK branch

graeme committed:
Point to latest commit (again)

graeme committed:
Add UserText for status view so far

graeme committed:
Use status observer and simplify tunnel controller

graeme committed:
Add to-design header view to status view

graeme committed:
Point to BSK 73.0.0

graeme committed:
Merge remote-tracking branch 'origin/develop' into graeme/netp-status-view

graeme commented on DuckDuckGo/NetworkProtectionStatusView.swift:
```diff
@@ -74,6 +98,49 @@ struct NetworkProtectionStatusView: View {
 
+private extension View {
+    @ViewBuilder
+    func hideScrollContentBackground() -> some View {
```
This is to override the system list background colour. Quite an ugly solution pre-iOS 16 but not sure how else to achieve it...

diegoreymendez commented on DuckDuckGo/NetworkProtectionStatusViewModel.swift:
```diff
@@ -21,59 +21,60 @@
+        self.headerTitle = Self.titleText(connected: statusObserver.recentValue.isConnected)
+        self.statusImageID = Self.statusImageID(connected: statusObserver.recentValue.isConnected)
+
+        isConnectedPublisher
+            .assign(to: \.isNetPEnabled, onWeaklyHeld: self)
+            .store(in: &cancellables)
+        isConnectedPublisher
+            .map(Self.titleText(connected:))
+            .assign(to: \.headerTitle, onWeaklyHeld: self)
+            .store(in: &cancellables)
+        isConnectedPublisher
+            .map(Self.statusImageID(connected:))
+            .assign(to: \.statusImageID, onWeaklyHeld: self)
+            .store(in: &cancellables)
```
Just a note that these publishers are not doing `.receive(on: DispatchQueue.main)`, whereas the ones below are.

Since the model is mostly used for UI we may want to add that for peace of mind.

graeme replied:
The shared publisher that these all use already dispatches onto main :) Just added it there to reduce the repetition

diegoreymendez commented on DuckDuckGo/NetworkProtectionStatusView.swift:
```diff
@@ -48,17 +57,32 @@ struct NetworkProtectionStatusView: View {
+                        .font(.system(size: 17, weight: .semibold))
+                        .multilineTextAlignment(.center)
+                        .foregroundColor(.titleText)
+                    Text(UserText.netPStatusHeaderMessage)
```
I'm sharing this with you because I didn't know this until yesterday, and because I have the doubt that the font is correct.  I'm not sure if `.system(...)` returns "SF Pro", so apologies if it does.

You can get the SwiftUI styling code directly from Figma, as shown below.

<img width="964" alt="Screenshot 2023-08-10 at 14 06 52" src="https://github.com/duckduckgo/iOS/assets/1836005/21fb15fe-9e7b-4f13-8c5a-28b0080b4283">

graeme replied:
Yeah, system does return that. But this is still useful as I did not know about that. Thanks!

diegoreymendez reviewed (commented):
I'm still testing but I'm also adding some comments for you to start looking at.

diegoreymendez commented on DuckDuckGo/NetworkProtectionStatusViewModel.swift:
```diff
@@ -86,20 +87,56 @@ final class NetworkProtectionStatusViewModel: ObservableObject {
+    var message: String {
+        switch self {
+        case .notConfigured:
+            return "Not Configured"
```
Just a note that at least in macOS we don't show this status even though it exists... we show `.disconnected`.

I don't think it needs to block this PR but something worth synching with between you, Robert and myself maybe?

diegoreymendez replied:
Also these strings should probably be `LocalizedString` (since they're using facing)?

I know SwiftUI handles that on its own, so I'm not sure whether this is needed or optional for iOS.  Please let me know so I can learn about it!

diegoreymendez commented on DuckDuckGo/NetworkProtectionStatusViewModel.swift:
```diff
@@ -86,20 +87,56 @@ final class NetworkProtectionStatusViewModel: ObservableObject {
+        case .notConfigured:
+            return "Not Configured"
+        case .disconnected:
+            return "Disconnected"
```
Adding to my comment above, in macOS the text for `.disconnected` is "Not connected".  I also checked the designs and it seems it's the same for iOS.

diegoreymendez reviewed (approved):
Added several comments to try and help with things that I believe may need to be adjusted (either in this PR or in a follow up).

But none of these things should be binding at this point, so I'm approving.

graeme committed:
NetP Status View toggle view (#1904)

Task/Issue URL: https://app.asana.com/0/0/1205232782620927/f

Description:

This is the second design implementation task in the overarching Network Protection Status View task and covers just the toggle view (that is, the view with the switch).

This is quite a Combine heavy one given the use of a Timer.publisher. I’m happy to explain this to anyone if it’s not clear.

This PR is stacked on #1899
